### PR TITLE
Explicitly define the branch to read from

### PR DESCRIPTION
**Pull request by Fydon**
Replacement for #4.

### CODE REVIEW

**Comment by vito** [comment]
This looks good to me, if it also looks good to @andreasf I'll go ahead and merge.

Thanks!

File: hg/hg.go [hunk 1]
```diff
@@ -204,10 +204,11 @@ func (self *Repository) Purge() (output []byte, err error) {
 }
 
 func (self *Repository) GetLatestCommitId() (output string, err error) {
+	branch := self.makeBranchQueryFragment()
 	include := self.makeIncludeQueryFragment()
 	exclude := self.makeExcludeQueryFragment()
 	tagFilter := self.maybeTagFilter()
-	revSet := fmt.Sprintf("last((((%s) - (%s)) & %s) - desc('[ci skip]'))", include, exclude, tagFilter)
+	revSet := fmt.Sprintf("last((((%s) - (%s)) & branch(%s) & %s) - desc('[ci skip]'))", include, exclude, branch, tagFilter)
```
Comment (andreasf) [inline hg/hg.go]: Good work!

The parameter to `branch()` should be quoted in single quotes, as branches can have parentheses in their names:

``` sh
$ hg branch ')'
marked working directory as branch )
```

You could then use `escapePath()` to escape quotes in branch names, which are also allowed:

``` sh
$ hg branch "'"
marked working directory as branch '
```

Adding `literal:` is a great idea. Instead of only adding it if the branch name starts with `re:`, I'd always add it, so that we don't need an extra function.

File: hg/hg_test.go [hunk 2]
```diff
@@ -27,6 +27,11 @@ var _ = Describe("Hg", func() {
 			Expect(unionOfPaths(repo.IncludePaths)).To(Equal("file('re:/path/1')|file('re:/path/2')|file('re:/path/3')"))
 		})
 
+		It("produces the correct branch fragment", func() {
+			Expect(repo.makeBranchQueryFragment()).To(Equal("literal:re:a_branch"))
+			Expect(emptyRepo.makeBranchQueryFragment()).To(Equal(""))
```
Comment (andreasf) [inline hg/hg_test.go]: If the user didn't explicitly specify a branch, we should probably assume `default`, not an empty string.

An empty unquoted string produces a parse error:

``` sh
$ hg log --rev "branch()"
hg: parse error: missing argument
```

And an empty quoted string returns results from the currently checked out branch:

``` sh
$ hg checkout bar-branch
1 files updated, 0 files merged, 0 files removed, 0 files unresolved
$ hg log --rev "branch('')"
changeset:   1:972cc008a6d4
branch:      bar-branch
user:        Foo
date:        Wed Nov 16 09:27:05 2016 +0100
summary:     bar

$ hg checkout foo-branch
2 files updated, 0 files merged, 0 files removed, 0 files unresolved
$ hg log --rev "branch('')"
changeset:   2:a649559327f9
branch:      foo-branch
tag:         tip
user:        Foo
date:        Wed Nov 16 09:27:26 2016 +0100
summary:     foo
```

File: test/test_check.sh [hunk 3]
```diff
@@ -74,7 +74,40 @@ test_it_skips_ignored_paths() {
...
+  # Test initial clone and that check doesn't see descendants that aren't in the given branch
+  local expected1=$(echo "[{\"ref\": $(echo $ref1 | jq -R .)}]"|jq ".")
+  assertEquals "$expected1" "$(check_uri $repo | jq '.')"
+
+  # Test check's first pull is still returning the same result, as above
+  local expected2=$(echo "[{\"ref\": $(echo $ref1 | jq -R .)}]"|jq ".")
+  assertEquals "$expected2" "$(check_uri $repo | jq '.')"
+
+  # Test check doesn't see changes to in other descendant branches when checking from a revision
+  local ref3=$(make_commit_to_branch $repo some-branch )
```
Comment (andreasf) [inline test/test_check.sh:93]: The commit `ref3` doesn't seem to change anything: `ref2` is already a commit on a branch different to `default`.

File: test/test_check.sh [hunk 4]
```diff
@@ -74,7 +74,40 @@ test_it_skips_ignored_paths() {
...
+  local expected3="[]"
+  assertEquals "$expected3" "$(check_uri_from $repo $ref1 | jq '.')"
+
+  # Test check only sees changes to its branch when checking from a revision
+  local ref4=$(make_commit $repo)
+
+  local expected4=$(echo "[{\"ref\": $(echo $ref4 | jq -R .)}]"|jq ".")
+  assertEquals "$expected4" "$(check_uri_from $repo $ref1 | jq '.')"
+
+  # Test check doesn't see changes to in parent branches
```
Comment (andreasf) [inline test/test_check.sh:104]: Mercurial doesn't have "parent branches", just parent commits.

A branch is the series of commits leading to a leaf commit (with no children itself). New commits are always appended to those leaf commits at the heads of branches, making them the new heads. A new commit therefore only appears in two branches if those branches are actually the same.

Because there is no distinction between "parent" and "child branches", we only need to make sure that we get results from the _expected_ branch.

**Review by andreasf — APPROVED** [review]

File: hg/hg_test.go (hunk 2 above)
Comment (Fydon) [reply hg/hg_test.go]: `default` is currently set in [check][2], [in][1] and [out][3], so it wouldn't be empty string when it isn't defined. I didn't think that there was a need to alter the test data, which was against an empty `Repository`. I could make that change that however I've removed `makeBranchQueryFragment` as requested and so there isn't a need for this test anymore.

[1]: https://github.com/concourse/hg-resource/blob/master/hgresource/in.go#L30
[2]: https://github.com/concourse/hg-resource/blob/master/hgresource/check.go#L28
[3]: https://github.com/concourse/hg-resource/blob/master/hgresource/out.go#L210

File: test/test_check.sh (hunk 3 above)
Comment (Fydon) [reply test/test_check.sh:93]: Right. Keep lines [95-96][1]? Remove this line or move it down to line [105][2] for `expected5`?

[1]: https://github.com/Fydon/hg-resource/blob/f56e39c02438610bc13c9c33183486f4f5b38c95/test/test_check.sh#L95-L96
[2]: https://github.com/Fydon/hg-resource/blob/f56e39c02438610bc13c9c33183486f4f5b38c95/test/test_check.sh#L105

File: hg/hg.go (hunk 1 above)
Comment (Fydon) [reply hg/hg.go]: Okay. Should I add parentheses and quotes to the name of the branches in `test_it_checks_correct_branch`?

File: test/test_check.sh (hunk 4 above)
Comment (Fydon) [reply test/test_check.sh:104]: True. Should I change the comments that refer to "parent branches" to say "parent commits in another branch" or rather remove the tests as well (lines [103-110][1]).

[1]: https://github.com/Fydon/hg-resource/blob/f56e39c02438610bc13c9c33183486f4f5b38c95/test/test_check.sh#L103-L110

**Comment by pivotal-issuemaster** [comment]
@Fydon Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal)!

[Click here](https://cla.pivotal.io/sync/pivotal?repositoryId=concourse/hg-resource&pullRequestId=5) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/faq) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@Fydon Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=concourse/hg-resource&pullRequestId=5)!

**Comment by Fydon** [comment]
@andreasf I've made some of the changes. Could please respond to my questions and check that the changes I've made are what you wanted?

**Comment by Fydon** [comment]
@andreasf Please respond

**Review by andreasf — APPROVED** [review]
Looks good to me, please merge

**Comment by andreasf** [comment]
Sorry for the late response. I'm not working on Concourse anymore and don't have commit rights to this repository. Thanks for your contribution!

**Comment by vito** [comment]
thanks!
